lib/datasets.go

```diff
@@ -489,88 +490,85 @@ func (m *DatasetMethods) Save(p *SaveParams, res *reporef.DatasetRef) error {
 		return fmt.Errorf("option to make dataset private not yet implemented, refer to https://github.com/qri-io/qri/issues/291 for updates")
 	}
 
-	ref, err := dsref.ParseHumanFriendly(p.Ref)
-	if errors.Is(err, dsref.ErrBadCaseName) {
-		// If dataset name is using bad-case characters, and is not yet in use, fail with error.
-		if !m.nameIsInUse(ctx, ref) {
+	// If the dscache doesn't exist yet, it will only be created if the appropriate flag enables it.
+	if p.UseDscache && !p.DryRun {
+		c := m.inst.dscache
+		c.CreateNewEnabled = true
+	}
+
+	// start with dataset fields provided by params
+	ds := p.Dataset
+	if ds == nil {
+		ds = &dataset.Dataset{}
+	}
+	ds.Assign(&dataset.Dataset{
+		BodyPath: p.BodyPath,
+		Commit: &dataset.Commit{
+			Title:   p.Title,
+			Message: p.Message,
+		},
+	})
+
+	if len(p.FilePaths) > 0 {
+		// TODO (b5): handle this with a qfs.Filesystem
+		dsf, err := ReadDatasetFiles(p.FilePaths...)
+		if err != nil {
 			return err
 		}
-		// If dataset name already exists, just log a warning and then continue.
-		log.Error(dsref.ErrBadCaseShouldRename)
-	} else if errors.Is(err, dsref.ErrEmptyRef) {
-		// Okay if reference is empty. Later code will try to infer the name from other parameters.
-	} else if errors.Is(err, dsref.ErrNotHumanFriendly) {
+		dsf.Assign(ds)
+		ds = dsf
+	}
+
+	if p.Ref == "" && ds.Name != "" {
+		p.Ref = fmt.Sprintf("me/%s", ds.Name)
+	}
+
+	resolver, err := m.inst.resolverForMode("local")
+	if err != nil {
 		return err
-	} else if err != nil {
-		// If some other parse error happened, describe a valid dataset name.
-		return dsref.ErrDescribeValidName
 	}
 
-	// Validate that username is our own, it's not valid to try to save a dataset with someone
-	// else's username. Without this check, base will replace the username with our own regardless,
-	// it's better to have an error to display, rather than silently ignore it.
 	pro, err := m.inst.repo.Profile()
 	if err != nil {
 		return err
 	}
-	if ref.Username != "" && ref.Username != "me" && ref.Username != pro.Peername {
-		return fmt.Errorf("cannot save using a different username than \"%s\"", pro.Peername)
-	}
 
-	// Parsed human-friendly dsref can only have username and name.
-	datasetRef := reporef.DatasetRef{
-		Peername: ref.Username,
-		Name:     ref.Name,
+	ref, isNew, err := base.PrepareSaveRef(ctx, pro, m.inst.logbook, resolver, p.Ref, ds.BodyPath, p.NewName)
+	if err != nil {
+		return err
 	}
 
-	ds := &dataset.Dataset{}
+	success := false
+	rollback := func() {
+		// if creating a new dataset fails, or we're doing a dry-run on a new dataset
+		// we need to remove the dataset
+		if isNew && (!success || p.DryRun) {
+			log.Debugf("removing unused log for new dataset %s", ref)
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second*2)
+			if err := m.inst.logbook.RemoveLog(ctx, ref); err != nil {
+				log.Errorf("couldn't cleanup unused reference: %q", err)
+			}
+			cancel()
+		}
+	}
+	defer rollback()
```

Comment:
This is a potential bug. I realize you're not assigning to `rollback` in this function, but it'll be easy for someone in the future to make the mistake of thinking that modifying `rollback` will change how it acts. However, that wouldn't work, see https://gist.github.com/dustmop/f91e66692f4e874331bdab0d5b339602

I would suggest either getting rid of the `rollback` variable and just calling `defer func() ...` directly, relying on the `success` check inside the function, or just changing this call to:

```
defer func() {
  rollback()
}()
```

Reply:
very good point. dropping the variable assignment in favour of a direct `defer func() { ...` call